### PR TITLE
Use an existing vpc

### LINKED ISSUE #29: Support for deploying clusters into an existing AWS VPC
/kind feature

Users should be able to deploy clusters into an existing VPC they have previously created.

Basic plan:

  - [ ] Create a new terraform stack named `network-existing-vpc` that takes vpc ID and subnet IDs as input
  - [ ] Run this stack instead of `network` if `existing-vpc/vpc-id` and `existing-vpc/subnet-ids` annotations are attached to a Cluster object
  - [ ] Some basic terraform to help test existing VPCs

### PR DESCRIPTION
**Pull request by kragniz**
**What this PR does / why we need it**:

This adds support for deploying into an existing VPC.

Enable by adding annotations to `network` in your tarmak.yaml like:

```yaml
  network:
    cidr: 10.99.0.0/16
    metadata:
      creationTimestamp: null
      annotations:
        tarmak.io/existing-vpc-id: vpc-xxxxxxxx
        tarmak.io/existing-public-subnet-ids: subnet-xxxxxxxx,subnet-xxxxxxxx
        tarmak.io/existing-private-subnet-ids: subnet-xxxxxxxx,subnet-xxxxxxxx
```

```release-note
Add experimental support for deploying clusters into existing AWS VPCs
```

Fixes #29.

### CODE REVIEW

File: terraform/amazon/network-existing-vpc/igw.tf [hunk 1]
```diff
@@ -0,0 +1,10 @@
+resource "aws_internet_gateway" "main" {
```
Comment (simonswine) [inline terraform/amazon/network-existing-vpc/igw.tf]: I think you can only have a single IGW per VPC, so you shouldn't create a new one

File: terraform/amazon/network-existing-vpc/vpc_endpoints.tf [hunk 2]
```diff
@@ -0,0 +1,5 @@
+resource "aws_vpc_endpoint" "s3" {
+  vpc_id          = "${var.vpc_id}"
+  service_name    = "com.amazonaws.${var.region}.s3"
+  route_table_ids = ["${data.aws_route_table.private.*.id}"]
```
Comment (simonswine) [inline terraform/amazon/network-existing-vpc/vpc_endpoints.tf]: I think a distinct makes sense here: 
https://www.terraform.io/docs/configuration/interpolation.html#distinct-list-

File: terraform/amazon/network-existing-vpc/vpc_peering.tf [hunk 3]
```diff
@@ -0,0 +1,59 @@
+data "terraform_remote_state" "vpc_peer_stack" {
```
Comment (simonswine) [inline terraform/amazon/network-existing-vpc/vpc_peering.tf]: I think this file isn't necessary, all existing vpc cluster need to be in the same VPC, so no peering required

File: terraform/amazon/network-existing-vpc/vpc_peering.tf [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+resource "aws_route" "them_peering_private" {
+  count                     = "${signum(length(var.vpc_peer_stack))*length(var.availability_zones)}"
+  route_table_id            = "${data.terraform_remote_state.vpc_peer_stack.route_table_private_ids[count.index]}"
+  destination_cidr_block    = "${var.network}"
+  vpc_peering_connection_id = "${aws_vpc_peering_connection.peering.id}"
+}
+
+resource "aws_route53_zone_association" "hub_zone" {
```
Comment (simonswine) [inline terraform/amazon/network-existing-vpc/vpc_peering.tf]: This is the exception to my comment from above

File: terraform/amazon/network-existing-vpc/vpc_endpoints.tf (hunk 2 above)
Comment (kragniz) [reply terraform/amazon/network-existing-vpc/vpc_endpoints.tf]: Added

File: terraform/amazon/network-existing-vpc/igw.tf (hunk 1 above)
Comment (kragniz) [reply terraform/amazon/network-existing-vpc/igw.tf]: Removed this file

**Comment by kragniz** [comment]
Tested and successfully completed a build using this VPC:

```terraform
provider "aws" {}

data "aws_availability_zones" "available" {}

resource "aws_vpc" "main" {
  cidr_block = "10.0.0.0/16"
  enable_dns_support   = true
  enable_dns_hostnames = true

  tags {
    Name = "test_vpc"
  }
}

resource "aws_eip" "nat" {
  vpc = true
}

resource "aws_subnet" "public" {
  count             = "${length(data.aws_availability_zones.available.names)}"
  vpc_id            = "${aws_vpc.main.id}"
  cidr_block        = "${cidrsubnet(cidrsubnet(aws_vpc.main.cidr_block, 3, 0), 3, count.index)}"
  availability_zone = "${data.aws_availability_zones.available.names[count.index]}"

  tags {
    Name = "public_${data.aws_availability_zones.available.names[count.index]}"
  }
}

resource "aws_subnet" "private" {
  count             = "${length(data.aws_availability_zones.available.names)}"
  vpc_id            = "${aws_vpc.main.id}"
  cidr_block        = "${cidrsubnet(aws_vpc.main.cidr_block, 3, count.index + 1)}"
  availability_zone = "${data.aws_availability_zones.available.names[count.index]}"

  tags {
    Name = "private_${data.aws_availability_zones.available.names[count.index]}"
  }
}

resource "aws_internet_gateway" "main" {
  vpc_id = "${aws_vpc.main.id}"
}

resource "aws_nat_gateway" "main" {
  depends_on    = ["aws_internet_gateway.main"]
  allocation_id = "${aws_eip.nat.id}"
  subnet_id     = "${aws_subnet.public.1.id}"
}

resource "aws_route_table" "public" {
  vpc_id = "${aws_vpc.main.id}"
}

resource "aws_route" "public" {
  route_table_id         = "${aws_route_table.public.id}"
  destination_cidr_block = "0.0.0.0/0"
  gateway_id             = "${aws_internet_gateway.main.id}"
}

resource "aws_route_table" "private" {
  vpc_id = "${aws_vpc.main.id}"
}

resource "aws_route" "private" {
  route_table_id         = "${aws_route_table.private.id}"
  destination_cidr_block = "0.0.0.0/0"
  nat_gateway_id         = "${aws_nat_gateway.main.id}"
}

resource "aws_route_table_association" "public" {
  count          = "${length(data.aws_availability_zones.available.names)}"
  subnet_id      = "${aws_subnet.public.*.id[count.index]}"
  route_table_id = "${aws_route_table.public.id}"
}

resource "aws_route_table_association" "private" {
  count          = "${length(data.aws_availability_zones.available.names)}"
  subnet_id      = "${aws_subnet.private.*.id[count.index]}"
  route_table_id = "${aws_route_table.private.id}"
}
```

**Comment by simonswine** [comment]
@kragniz can we get that example terraform into a doc and some instructions (with a big warning that this is experimental), otherwise looking good to me

**Comment by kragniz** [comment]
I added some basic docs and a release note.

**Comment by simonswine** [comment]
Thanks a lot for that @kragniz 

/approve
/lgtm

**Pull request merged**